### ISSUE TITLE
ICE when using declarative macro being defined by another declarative macro

### ISSUE THREAD

**Issue by mwu-tow** [rust-lang/rust-clippy#7272]
### Code
Reduced code is at https://github.com/mwu-tow/clippy-ice-repro

Run the nightly clippy on this repo to reproduce the issue,

It consists of `lib.rs` and `main.rs`.  Lib defines a macro with a macro (using a trick where first argument expands to `$` sign in the nested macro), `main.rs` uses it.

<details><summary>lib.rs</summary>
  <p>

```rust
pub fn warn<T>(_:T) {}

macro_rules! define_macro {
    ($d:tt $lower:ident $upper:ident) => {
        #[macro_export] macro_rules! $upper  {
            ($d arg:tt) => { $crate:: $lower($d arg) }
        }
    }
}

define_macro!{$ warn  WARNING}
```
  </p>
</details>

<details><summary>main.rs</summary>
  <p>

```rust
use repro::*;

pub fn main() { || WARNING!("Style changed!"); }
```
  </p>
</details>

I haven't been able to reduce this to use only a single cargo target.

This is a regression from stable 1.52.1. 
The `nightly-2021-03-25-x86_64-pc-windows-gnu` is confirmed to work correctly.
The `nightly-2021-04-13-x86_64-pc-windows-gnu` is confirmed to ICE on this repo.

### Meta

- `cargo clippy -V`: clippy 0.1.54 (79e50bf7 2021-05-10)
- `rustc -Vv`:
```
rustc 1.54.0-nightly (79e50bf77 2021-05-10)
binary: rustc
commit-hash: 79e50bf77928f374921a6bcafee3fcff1915f062
commit-date: 2021-05-10
host: x86_64-pc-windows-gnu
release: 1.54.0-nightly
LLVM version: 12.0.1
```

### Error output

```
❯ cargo clippy
    Checking repro v0.1.0 (H:\NBO\clippy-ice-repro)
warning: statement with no effect
 --> src\main.rs:3:17
  |
3 | pub fn main() { || WARNING!("Style changed!"); }
  |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(clippy::no_effect)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#no_effect

thread 'rustc' panicked at 'begin <= end (59 <= 42) when slicing `$crate:: $lower($d arg) }
        }
    }
}

define_macro!{$ warn`', src\tools\clippy\clippy_lints\src\functions\too_many_lines.rs:24:26


error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust-clippy/issues/new

note: Clippy version: clippy 0.1.54 (79e50bf7 2021-05-10)

query stack during panic:
#0 [analysis] running analysis passes on this crate
end of query stack
warning: 1 warning emitted

error: could not compile `repro`
```

<details><summary>Backtrace</summary>
  <p>
  
  ```
stack backtrace:
   0: rust_begin_unwind
             at /rustc/79e50bf77928f374921a6bcafee3fcff1915f062\/library\std\src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/79e50bf77928f374921a6bcafee3fcff1915f062\/library\core\src/panicking.rs:92:14
   2: core::str::slice_error_fail
   3: clippy_lints::functions::too_many_lines::check_fn
   4: <clippy_lints::functions::Functions as rustc_lint::passes::LateLintPass>::check_fn
   5: <rustc_lint::late::LateLintPassObjects as rustc_lint::passes::LateLintPass>::check_fn
   6: rustc_hir::intravisit::walk_expr
   7: rustc_hir::intravisit::walk_stmt
   8: rustc_hir::intravisit::walk_block
   9: rustc_hir::intravisit::walk_expr
  10: rustc_hir::intravisit::walk_body
  11: <rustc_lint::late::LateContextAndPass<T> as rustc_hir::intravisit::Visitor>::visit_nested_body
  12: <rustc_lint::late::LateContextAndPass<T> as rustc_hir::intravisit::Visitor>::visit_fn
  13: rustc_hir::intravisit::walk_item
  14: rustc_hir::intravisit::Visitor::visit_nested_item
  15: rustc_hir::intravisit::walk_crate
  16: rustc_lint::late::late_lint_pass_crate
  17: rustc_lint::late::late_lint_crate
  18: rustc_session::utils::<impl rustc_session::session::Session>::time
  19: rustc_interface::passes::analysis::{{closure}}::{{closure}}
  20: rustc_interface::passes::analysis
  21: rustc_middle::dep_graph::<impl rustc_query_system::dep_graph::DepKind for rustc_middle::dep_graph::dep_node::DepKind>::with_deps
  22: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  23: rustc_data_structures::stack::ensure_sufficient_stack
  24: rustc_query_system::query::plumbing::force_query_with_job
  25: rustc_query_system::query::plumbing::get_query_impl
  26: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::analysis
  27: rustc_interface::passes::QueryContext::enter
  28: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  29: rustc_span::with_source_map
  30: rustc_interface::interface::create_compiler_and_run
  31: scoped_tls::ScopedKey<T>::set
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
  ```
  
  </p>
</details>
